### DISCUSSION TITLE
TypeError: Cannot read property 'tap' of undefined - @next/react-refresh-utils

### DISCUSSION THREAD

**Discussion by gregegan** [vercel/next.js#19081] [Help]
# Bug report

## Describe the bug

While trying to upgrade `@apollo/react-hooks` to `@apollo/client` it appears that I am now getting issues on `Cannot read property 'tap' of undefined`. This error occurs when running `next dev`

```
TypeError: Cannot read property 'tap' of undefined
    at /Users/gregegan/repos/the-landing/client/node_modules/@next/react-refresh-utils/ReactRefreshWebpackPlugin.js:100:61
    at SyncHook.eval [as call] (eval at create (/Users/gregegan/repos/the-landing/client/node_modules/next/node_modules/tapable/lib/HookCodeFactory.js:19:10), <anonymous>:7:1)
```

## Screenshots

![image](https://user-images.githubusercontent.com/6665100/98877101-90225300-2434-11eb-9f5a-64b045d6de74.png)

```
gregegan@Gregs-MacBook-Pro client % npm run dev

> the-landing@0.1.0 dev
> next dev

Loaded env from /Users/gregegan/repos/the-landing/client/.env
ready - started server on http://localhost:3000
TypeError: Cannot read property 'tap' of undefined
    at /Users/gregegan/repos/the-landing/client/node_modules/@next/react-refresh-utils/ReactRefreshWebpackPlugin.js:100:61
    at SyncHook.eval [as call] (eval at create (/Users/gregegan/repos/the-landing/client/node_modules/next/node_modules/tapable/lib/HookCodeFactory.js:19:10), <anonymous>:7:1)
    at SyncHook.lazyCompileHook (/Users/gregegan/repos/the-landing/client/node_modules/next/node_modules/tapable/lib/Hook.js:154:20)
    at Compiler.newCompilation (/Users/gregegan/repos/the-landing/client/node_modules/next/node_modules/webpack/lib/Compiler.js:631:26)
    at /Users/gregegan/repos/the-landing/client/node_modules/next/node_modules/webpack/lib/Compiler.js:667:29
    at AsyncSeriesHook.eval [as callAsync] (eval at create (/Users/gregegan/repos/the-landing/client/node_modules/next/node_modules/tapable/lib/HookCodeFactory.js:33:10), <anonymous>:6:1)
    at AsyncSeriesHook.lazyCompileHook (/Users/gregegan/repos/the-landing/client/node_modules/next/node_modules/tapable/lib/Hook.js:154:20)
    at Compiler.compile (/Users/gregegan/repos/the-landing/client/node_modules/next/node_modules/webpack/lib/Compiler.js:662:28)
    at /Users/gregegan/repos/the-landing/client/node_modules/next/node_modules/webpack/lib/Watching.js:77:18
    at AsyncSeriesHook.eval [as callAsync] (eval at create (/Users/gregegan/repos/the-landing/client/node_modules/next/node_modules/tapable/lib/HookCodeFactory.js:33:10), <anonymous>:15:1)
npm ERR! code 1
npm ERR! path /Users/gregegan/repos/the-landing/client
npm ERR! command failed
npm ERR! command sh -c next dev

npm ERR! A complete log of this run can be found in:
npm ERR!     /Users/gregegan/.npm/_logs/2020-11-11T23_40_58_534Z-debug.log
```

## System information

- OS: Mac
- Version of Next.js: [e.g. 10.0.1]
- Version of Node.js: [e.g. 15.0.1]

## Additional context

Here is the package.json
https://gist.github.com/gregegan/91793416b5b2b70e17501497ee9a2839

Here is the package-lock.json
https://gist.github.com/gregegan/acbdf34718350b630dc30dbb4e9a3bfc

**Comment by chunghsien** [comment]
#Next.js v10.0.1 with webpack v5.4.0 bugs

$ cd <project_folder>node_modules/next/node_modules

#Modify package.json file`
#Replace "webpack-sources": "1.4.3" => "webpack-sources": "^1.4.3"
#Replace "pnp-webpack-plugin": "1.6.4" => "pnp-webpack-plugin": "^1.6.4"
#Replace "webpack": "4.44.2" => "webpack": "^4.44.2"
#Replace "webpack-sources": "1.4.3" => "webpack-sources": "^1.4.3"

$ npm update`

**Reply by hansfpc** [reply]
Thanks! worked for me :D

**Reply by RiansyahTohamba** [reply]
thanks! worked for me too :)

**Comment by gregegan** [comment]
Thanks @chunghsien - Should this fix be going into the next.js repo rather then me forking it and updating it manually?

**Comment by chunghsien** [comment]
Yes, this error is because Next.js v10.0.1 uses webpack 4.x instead of version 5.

**Comment by gregegan** [comment]
Thank you @chunghsien.

Is someone already addressing this? Is this a known issue? Seems like it would be critical for @apollo/client to work with nextjs

**Comment by gregegan** [comment]
@chunghsien - does nextjs v10.0.1 have plans to support the latest webpack? Or should I stop trying to upgrade to the latest @apollo/client

**Comment by chunghsien** [comment]
I suggest using webpack 4.44.2.

![未命名-1](https://user-images.githubusercontent.com/1892664/99265697-eb04e180-285c-11eb-8795-7b7783c65717.png)

**Comment by timneutkens** [comment]
Converted to discussions given that the issue template was not followed (no reproduction provided).

**Comment by arvigeus** [comment]
This bit me as well. It was working fine yesterday.

Repo to reproduce: https://github.com/arvigeus/react-modern-state

npm v7

**Comment by hmedney** [comment]
In case this helps anyone, I encountered this issue using `webpack@5.12.3` and `next@10.0.5` when installed via `npm@6.14.7`. When I installed the same `package.json` via `yarn@1.22.4`, the issue disappeared.

**Reply by saqbach** [reply]
Wild. I have absolutely no idea why that worked, but it did.

**Reply by coryodaniel** [reply]
Same here.

**Reply by ctsstc** [reply]
Maybe you had a package-lock and not a yarn lock file?  Potentially there are no locks in place if you're moving over to yarn or that the locked versions do not match. Potentially it would be like deleting your lock file and running `npm install` again.

**Comment by ManzoliW** [comment]
I installed with yarn and and the issue disappeared, but to use with node i have to use the 4.44.2 version of webpack.

$npm install --save-dev webpack@4.44.2

**Reply by dfoverdx** [reply]
This was exactly my issue, but of course I didn't see this comment until I had figured it out myself.  🤦‍♀️

For posterity, I'm going to add the words: `create-react-app` and `cra`.  Maybe Google will pick it up and include the proper hash.  ...Not that I've ever seen Google links use a hash, but who knows.  🤷‍♀️

I'll also say that I have _no_ idea why using yarn instead of npm makes a difference.
